File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            if (isVowel(s[left]) and isVowel(s[right])) {
                swap(s[left], s[right]);
                left++;
                right--;
            }
            
            while (left < s.length() and !isVowel(s[left])) {
                left++;
            }
            
            while (right > -1 and !isVowel(s[right])) {
                right--;
            }
        }

        return s;
    }

    bool isVowel(char c) {
        switch (c) {
            case 'a':
                return true;
            case 'A':
                return true;
            case 'e':
                return true;
            case 'E':
                return true;
            case 'i':
                return true;
            case 'I':
                return true;
            case 'o':
                return true;
            case 'O':
                return true;
            case 'u':
                return true;
            case 'U':
                return true;
            default:
                return false;
        }
    }
};

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.java
class Solution {
    public String reverseVowels(String s) {
        StringBuilder word = new StringBuilder(s);
        int left = 0, right = s.length() - 1;
        boolean leftFound = false, rightFound = false;

        while (left < right) {
            if (isVowel(s.charAt(left)))  {
                leftFound = true;
            }

            if (isVowel(s.charAt(right))) {
                rightFound = true;
            }

            if (leftFound && rightFound) {
                word.setCharAt(left, s.charAt(right));
                word.setCharAt(right, s.charAt(left));

                leftFound = false;
                rightFound = false;
            }

            if (!leftFound) {
                left++;
            }

            if (!rightFound) {
                right--;
            }
        }

        return word.toString();
    }

    private boolean isVowel(char c) {
        return (c == 'a' || c == 'A' || c == 'e' || c == 'E' || c == 'i' || c == 'I' || c == 'o' || c == 'O' || c == 'u' || c == 'U');
    }
}